src/lib.rs

```diff
@@ -364,18 +364,67 @@ impl Rng {
         }
     }
 
+    /// Generates a random `f32` in range `0..=1`.
+    #[inline]
+    pub fn f32_inclusive(&mut self) -> f32 {
+        // Generate a number in 0..2^63 then convert to f32 and multiply by 2^(-63).
+        //
+        // Even though we're returning f32, we still generate u64 internally to make
+        // it possible to return nonzero numbers as small as 2^(-63). If we only
+        // generated u32 internally, the smallest nonzero number we could return
+        // would be 2^(-32).
+        //
+        // The integer we generate is in 0..2^63 rather than 0..2^64 to improve speed
+        // on x86-64, which has efficient i64->float conversion (cvtsi2ss) but for
+        // which u64->float conversion must be implemented in software.
+        //
+        // There is still some remaining bias in the int-to-float conversion, because
+        // nonzero numbers <=2^(-64) are never generated, even though they are
+        // expressible in f32. However, at this point the bias in int-to-float conversion
+        // is no larger than the bias in the underlying WyRand generator: since it only
+        // has a 64-bit state, it necessarily already have biases of at least 2^(-64)
+        // probability.
+        //
+        // See e.g. Section 3.1 of Thomas, David B., et al. "Gaussian random number generators,
+        // https://www.doc.ic.ac.uk/~wl/papers/07/csur07dt.pdf, for background.
+        const MUL: f32 = 1.0 / (1u64 << 63) as f32;
+        (self.gen_u64() >> 1) as f32 * MUL
+    }
+
     /// Generates a random `f32` in range `0..1`.
+    ///
+    /// Function `f32_inclusive()` is a little simpler and faster, so default
+    /// to that if inclusive range is acceptable.
+    #[inline]
     pub fn f32(&mut self) -> f32 {
-        let b = 32;
-        let f = core::f32::MANTISSA_DIGITS - 1;
-        f32::from_bits((1 << (b - 2)) - (1 << f) + (self.u32(..) >> (b - f))) - 1.0
+        loop {
+            let x = self.f32_inclusive();
+            if x < 1.0 {
+                return x;
+            }
+        }
+    }
```

Comment:
Was the perf of this measured? Seems like adding a branch and a loop to `f32` might have performance implications I'm not comfortable with releasing.

Reply:
(These numbers are the benchmark numbers, divided by 10_000, since the benchmark itself generates 10_000 f32s.)